DRA kubelet: connection monitoring

pohly opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

When a kubelet plugin shuts down without removing its registration socket (for example, because it was forcibly killed or crashed), then the kubelet does not clean up and delete ResourceSlices even if the plugin does not come back. With connection monitoring of the service socket, the kubelet recognizes when a plugin becomes unusable and then cleans up accordingly.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/128696
Replaces https://github.com/kubernetes/kubernetes/pull/131073
Depends on https://github.com/kubernetes/kubernetes/pull/132096

#### Special notes for your reviewer:

Please review commit-by-commit. The initial commits refactor code to make the last commits simpler, so at least https://github.com/kubernetes/kubernetes/pull/132096 should get merged first.

#### Does this PR introduce a user-facing change?

```release-note
DRA kubelet: the kubelet now also cleans up ResourceSlices in some additional failure scenarios (driver gets removed forcibly or crashes and does not restart).
```

bart0sh committed:
DRA kubelet: add connection monitoring

This ensures that ResourceSlices get removed also when a plugin becomes
unresponsive without removing the registration socket.

Tests are from https://github.com/kubernetes/kubernetes/pull/131073 by Ed
with some modifications, the implementation is new.

pohly committed:
DRA kubelet: simplify plugin creation and connection handling

Instead of creating the gRPC connection on demand and forcing gRPC to connect,
we establish it immediately and rely on gRPC to handle the underlying
connection automatically like it usually does.

It's not clear what benefit the one second connection timeout had. The way it
is now, gRPC calls still fail when the underlying connection cannot be
established. Having to have a separate context for establishing that connection
just made the code more complex.

The DRAPluginManager is the central component which manages plugins. Making it responsible
for creating them reduces the number of places where a DRAPlugin struct needs to
be initialized. Doing this in the DRAPluginManager instead of a stand-alone function
simplifies the implementation of connection monitoring, because that will be
something that is tied to the DRAPluginManager state.

pohly committed:
DRA kubelet: use TimedWorkersQueue

Conceptually TimedWorkersQueue is similar to the current code: it spawns
goroutines and cancels them. Using it makes the code a bit shorter, even though
the TimedWorkersQueue API could be a bit nicer and more consistent (key string
vs. WorkerArgs as parameters).

Depending on the tainteviction package is a bit odd, which is the reason why
TimedWorkersQueue wasn't already used earlier. But there don't seem to be other
implementations of this common
problem. https://pkg.go.dev/k8s.io/client-go/util/workqueue#TypedDelayingInterface
doesn't work because queue entries cannot be removed.

This doesn't really solve the problem of tracking goroutines for wiping because
TimedWorkersQueue doesn't support that. But not tracking is arguably better
than doing it wrong and this only affects unit tests, so it should be okay.

pohly commented:
/test pull-kubernetes-verify pull-kubernetes-unit pull-kubernetes-integration pull-kubernetes-node-e2e-crio-cgrpv1-dra pull-kubernetes-kind-dra-all pull-kubernetes-node-e2e-containerd-2-0-dra

bart0sh commented on pkg/kubelet/cm/dra/plugin/plugins_store.go:
```diff
@@ -17,80 +17,394 @@ limitations under the License.
+	default:
+		return
+	}
+	if m.store.backgroundCtx.Err() != nil {
```
Should the store be locked before accessing its attributes?

bart0sh commented on pkg/kubelet/pluginmanager/pluginwatcher/README.md:
```diff
@@ -22,6 +22,58 @@ should end with a DNS domain that is unique for the plugin. Each time a plugin
 starts, it has to delete old sockets if they exist and listen anew under the
 same filename.
 
+## Monitoring Plugin Connection
```
As this is fully implemented in the dra code, would it make sense to move the documentation there? Monitoring service connection is out of scope of `pluginwatcher`, I believe.
Moving it out of `pluginmanager` code would also simplify the approval of this PR as it can be approved only by DRA approvers.

pohly replied:
We also need SIG Node approval for the DRA code, moving the documentation won't simplify approval. Besides, @klueska can approve also this here.

I prefer to keep documentation about kubelet plugin implementation aspects in one place. Perhaps other types will also implement this.

klueska commented on pkg/kubelet/cm/dra/plugin/plugin_manager.go:
```diff
@@ -0,0 +1,503 @@
+//
+// It returns an informative error message including the driver name
+// with an explanation why the driver is not usable.
+func (pm *PluginManager) GetDRAPlugin(driverName string) (*Plugin, error) {
```
Does it make sense to rename this type to `DRAPluginManager`, the returned type to `DRAPlugin` and then have this function just be `called GetPlugin()`?

pohly replied:
Looks good to me. I'll update the renaming commit...

bart0sh commented on pkg/kubelet/cm/dra/plugin/plugins_store.go:
```diff
@@ -17,80 +17,394 @@ limitations under the License.
+	//
+	// TimedWorkerQueue uses namespace/name as key. We use
+	// the driver name as name with no namespace.
+	pendingWipes *timedworkers.TimedWorkerQueue
```
I didn't equate it to "Got object". I said it's a big step in that direction, which I still believe it is. Merging registration and connection management functionality into a storage object doesn't look like a good thing to do. Even if a result seemingly makes sense as you've mentioned above, it doesn't mean that descomposing registration, connection management or cleanup into a separate entities wouldn't make sense. In my opinion it would be easier to understand, maintain and test.

bart0sh commented on test/e2e_node/dra_test.go:
```diff
@@ -568,6 +588,48 @@ var _ = framework.SIGDescribe("node")(framework.WithLabel("DRA"), feature.Dynami
+			ginkgo.By("wait for ResourceSlice removal")
+			gomega.Eventually(ctx, listResources).Should(gomega.BeEmpty(), "ResourceSlices without plugin")
+			gomega.Consistently(ctx, listResources).WithTimeout(5*time.Second).Should(gomega.BeEmpty(), "ResourceSlices without plugin")
+		})
```
> we allow the slice to be deleted and have the restarted driver recreate it and see it remain, but that doesn't really tell us whether the kubelet has reconnected.

I thought about making sure that the slice(s) deleted by the Kubelet before the driver publishes them again on reconnect by:
- stopping the plugin
- waiting for slices to be removed by kubelet
- starting the plugin
- may be waiting for the slices to be updated( not sure if plugin updates them on its start unconditionally)
- testing that it serves requests successfully

bart0sh replied:
One more thing - [here](https://github.com/kubernetes/kubernetes/pull/131073#discussion_r2087121475) you've mentioned that we need to test use case(s) with the same registration and service sockets. Would it still make sense to do?

pohly replied:
Yes, worth adding. I had it on my backlog to go through that post before merging.

pohly commented:
/test pull-kubernetes-e2e-kind-dra-canary pull-kubernetes-e2e-kind-dra-n-1-canary

pohly commented:
/test pull-kubernetes-kind-dra-n-1-canary pull-kubernetes-kind-dra-n-2-canary pull-kubernetes-kind-dra-canary

pohly commented:
/test pull-kubernetes-kind-dra-n-1-canary pull-kubernetes-kind-dra-n-2-canary pull-kubernetes-kind-dra-canary

pohly commented:
/test pull-kubernetes-kind-dra-n-1-canary pull-kubernetes-kind-dra-n-2-canary

pohly commented:
/test pull-kubernetes-kind-dra-n-1-canary pull-kubernetes-kind-dra-n-2-canary

bart0sh committed:
kubelet: DRA: fix linter warnings

Fixed the following warnings:
dra_test.go:884:2: singleCaseSwitch: should rewrite switch statement to if statement (gocritic)
	switch podName {
	^
dra_test.go:686:4: SA4006: this value of kubeletPlugin is never used (staticcheck)
	kubeletPlugin = newDRAService(ctx, f.ClientSet, nodeName, driverName)
        ^

bart0sh committed:
e2e_node: DRA: pass socket path as a parameter

Added an ability to specify the socket path for the DRA gRPC
service in the e2e node tests.

The PluginSocket option is added to allow setting the name
of the socket inside the directory where the DRA driver
creates the socket for the DRA gRPC calls. This is used by
the kubelet to connect to the DRA plugin.

The newDRAService and newRegistrar functions are updated to
accept a socketPath parameter, which is used to configure
the PluginDataDirectoryPath and PluginSocket options for the
DRA plugin.

This change enables more flexible configuration of the DRA
plugin in e2e tests, allowing for testing with different
socket paths.

bart0sh committed:
e2e_node: DRA: add tests for different socket setups

Added tests to verify DRA functionality with 2 different socket
configurations:
- the same socket is used for the registration and the DRA service
- 2 separate sockets are used for the registration and the DRA service

Used table-driven ginkgo to avoid code duplication:
specs https://onsi.github.io/ginkgo/#table-driven-tests

This change enhances the robustness of the DRA e2e tests by
validating its behavior with different socket setups.

bart0sh committed:
kubelet: DRA: fix test failure on Windows

Unix socket path has to be <= 108 characters in length on Windows.
Shortened DRA socket path for TestConnectionHandling and
TestRegistrationHandler tests should fix the test run on Windows.

bart0sh committed:
kubelet: DRA: fix unit test failure

Fixed race condition caused by removing already removed
socket directory.

bart0sh committed:
kubelet: DRA: fix TestRegistrationHandler

Using the same socket path for different test cases caused
test failure on windows:
listen unix C:\Users\azureuser\AppData\Local\Temp\TestRegistrationHandler3881105518\001/dra-plugin-a.sock:
bind: Only one usage of each socket address (protocol/network address/port) is normally permitted.

Creating unique socket path for every test case should fix it.

bart0sh commented:
/retest
/lgtm

/assign @klueska @SergeyKanzhelev 
for approval

pohly commented on pkg/kubelet/cm/dra/plugin/dra_plugin_manager.go:
```diff
@@ -58,21 +64,72 @@ type DRAPluginManager struct {
+	*DRAPlugin
+	pm *DRAPluginManager
+
+	// connected is protect by store.mutex.
```
Fixed, please re-add LGTM.
